### ISSUE TITLE
Abuse now won't load, with error message "unable to load cache file name c:\lcache00.tmp"

### ISSUE THREAD

**Issue by Gryzor1363** [dosbox-staging/dosbox-staging#4123]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.82

### What Operating System are you using?

Windows 7

### Relevant hardware info

FX-8350 processor, GTX 1070 graphics card, 16 Gb of RAM.

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Upon game launch, error appears as described in title.

Games still works using the same configuration as usual and confirmed as working on mainline Dosbox 0.74, and also Dosbox Staging 0.80 and 0.81. 

Also tried lowering CPU to 3000 Mips, RAM to 16, type to either 486 or Pentium, tweaked EMS/XMS/UMB, all point to the same error specifically in 0.82.

I have tried granting Dosbox full permissions and I/O and run it as administrator, set _**allow_write_protected_files = true**_ within the config, switched the game folder and its parent to read/write mode as well, as per recommendation on very old threads pertaining to a similar issue in the past with older versions of Dosbox, tried deleting the temporary file in question, to no avail. 

Please see output details from game in attachment. 

![image](https://github.com/user-attachments/assets/f0a4d254-38b0-4750-abea-0c9718b878cf)

### Steps to reproduce the behavior.

Explain how to reproduce
1. Launch game
2.Wait until loading hits a wall with the error message as described.

### Download URL of affected game or software

https://www.myabandonware.com/game/abuse-1ne

### Your configuration

```ini
https://gist.github.com/Gryzor1363/1c944cebcde5f077252f77972743589b
```

### Code of Conduct & Contributing Guidelines

- [x] Yes, I agree.

**Comment by Gryzor1363** [comment]
Thanks for reacting @weirddan455 . I cannot reproduce on Staging 0.80, 0.81 or mainline 0.74-3 either, all pointing to the same exact game folder and file locations on my computer, and none of them even need running as admin, so it's definitely not user or permissions-related on OS-level.
But considering your 0.82 binary works fine, I'm starting to think something related to Win 7 that I'm using, might still be the culprit... I'll try with the same exe and conf on another computer ASAP and come back with results.

**Comment by weirddan455** [comment]
The file creation failed error narrows the problem down.  It's failing here in the code:

https://github.com/dosbox-staging/dosbox-staging/blob/f7ae7dfdaa279ee434f2079b50aa2e562f366182/src/dos/drive_local.cpp#L81-L87

It could be failing for any number of reasons though, including permissions.

Can you see if you can create a file with a simple DOS command like `echo foo > test.txt`?  See if that creates the file or if it logs the same error.  Then try again mounting your C drive somewhere else.

**Comment by Gryzor1363** [comment]
The echo command indeed creates file with success. There is indeed a new test.txt file created with the "foo" string inside.
Still no luck running Abuse.exe though. 
Further investigated file creation by expressly saving in Prince of Persia, file is indeed updated with new data, and its folder has the same parent as Abuse.
This is getting tricky.

**Comment by weirddan455** [comment]
>   But considering your 0.82 binary works fine, I'm starting to think something related to Win 7 that I'm using, might still be the culprit... I'll try with the same exe and conf on another computer ASAP and come back with results.

Well, let me know if you can test on another computer.  I just checked the system requirements on our website and it says Windows 7 is only officially supported through v0.80.1:  https://www.dosbox-staging.org/releases/windows/

How did you get v0.82.0 to work?  The binary from dosbox-staging.org site just happened to work on Win 7 or did you get it from somewhere else?

**Comment by johnnovak** [comment]
@Gryzor1363 Yes, do note we only support Windows 10 64-bit and beyond now. Win 7 support (and testing) was dropped for quitesome time now.

But this seems like a permission related issue.

**Comment by Gryzor1363** [comment]
@weirddan455 yes I tried my luck on my existing Win 7 install, while I have plans to migrate to 10 in just a few weeks. I took the portable package from the official download page and it happens to luckily still work like a charm for every single game of my collection except, obviously, for this one.

@johnnovak I realize it might not necessarily be an appropriate channel to query for support on an obsolete OS... I admittedly allowed myself to do so considering the issue does as you say seem constrained to a non-OS specific issue, still, I would lie if I said I'm expert enough to know that for certain, still.

Does it mean I should close this and find a workaround away from that thread ? I will understand if no time and effort can be devoted to this. Thanks for considering and following up so far. Or maybe that kind of issue can appear on newer OSes and solving this can furnish insight for anybody in the future experiencing this, you will be the judges.

Thanks for the newest release anyway, does wonders with all of my other games ;)

**Comment by johnnovak** [comment]
@Gryzor1363 No prob, if no one can reproduce it we might chalk it up to Win7 being the culprit. But I think it's more filesystem permissions related.

Good to hear Staging still works on Win 7, though 😄 We simply don't have the ability to do testing on Win 7 anymore (e.g., my new motherboard I bought earlier this year only has Win 10 and 11 64-bit drivers).

**Comment by weirddan455** [comment]
It might not be a Windows 7 specific issue.  I wrote most of that Win32 file I/O code and I don't think it's using any newer APIs.  Honestly, it looks like file permissions issue even though you said you checked that.  I would try creating a new DOS root directory in your My Documents or something where you know your user has full read/write permissions and try again.  Just to rule that out.

**Comment by Gryzor1363** [comment]
Thank you both for following up then , truly appreciated ;)

@weirddan455 I just tried the my docs trick, and something interesting happened : issue is still present with the same in-game error message but the existing LCACHE00.TMP file gets erased, I guess "half" the rights are there. Still no "create", but "delete" seem allowed this time within the Home directory, which itself is quite odd as the permission within it for all users are strictly identical and provide full control for both my active user and of course admin...

Incredible that the whole issue impacts one and only folder though... every other game has its files deleted/created/modified with zero hiccups (namely savegames, etc... tested several).

**Comment by weirddan455** [comment]
@Gryzor1363 You said a lot of things there.  The relevant thing I'm hearing is that if you delete `LCACHE00.TMP` the game works?  And it successfully creates a new file?  If you then re-launch Dosbox after a successful game launch, does it continue to work or do you have to delete the file every time?

`create_native_file()` is supposed to work even if the file already exists.  In that case, it's supposed to truncate it to zero (essentially overwriting the file with a new copy).

**Comment by Gryzor1363** [comment]
Please let me clarify :

- Creating a brand new folder _C:\Games\Test_ in which I copied all the game files and trying to launch the game led to the very same error message pasted in my OP. 
- This time however, I noticed that Dosbox was able to delete the _LCACHE00.TMP_ file upon game launch, which it did not, initially, in my previous attempt in the game's existing folder (which was _C:\Games\Abuse_)
- From that point on, no matter how many times I try to launch the game, Dosbox never creates that file again, and game initialization continues to fail with the very same recurring error output as before.
- If I paste a spare copy of _LCACHE00.TMP_ in the _C:\Games\Test_ folder and launch the game again from there, it gets deleted again, still does not get created anew, and game still fails to load with the same error illustrated in the screencap.

So I can confirm I can reproduce that part of the new behavior (Dosbox now deleting the cache file every time I try running the game), but the issue remains the same in the end at this point (the file does not get recreated and game load fails).

I hope it was clearer this time.

**Comment by johnnovak** [comment]
> Please let me clarify :
> 
>     * Creating a brand new folder _C:\Games\Test_ in which I copied all the game files and trying to launch the game led to the very same error message pasted in my OP.
> 
>     * This time however, I noticed that Dosbox was able to delete the _LCACHE00.TMP_ file upon game launch, which it did not, initially, in my previous attempt in the game's existing folder (which was _C:\Games\Abuse_)
> 
>     * From that point on, no matter how many times I try to launch the game, Dosbox never creates that file again, and game initialization continues to fail with the very same recurring error output as before.
> 
>     * If I paste a spare copy of _LCACHE00.TMP_ in the _C:\Games\Test_ folder and launch the game again from there, it gets deleted again, still does not get created anew, and game still fails to load with the same error illustrated in the screencap.
> 
> 
> So I can confirm I can reproduce that part of the new behavior (Dosbox now deleting the cache file every time I try running the game), but the issue remains the same in the end at this point (the file does not get recreated and game load fails).
> 
> I hope it was clearer this time.

Try it again in `C:\BrandNewFolder` instead of `C:\Games` perhaps.

**Comment by weirddan455** [comment]
>  Try it again in C:\BrandNewFolder instead of C:\Games perhaps.

Yeah, it's the root that I was suspecting of having a permissions issue, not your Abuse game folder.  That's where the cache file gets created.  I would also try My Documents or somewhere in C:\Users\your-username\ as Windows sometimes gets weird with permissions outside of your user folder.

**Comment by Gryzor1363** [comment]
> Try it again in `C:\BrandNewFolder` instead of `C:\Games` perhaps.

I actually did that before as well, should have mentioned it, my bad. 
Tried from _C:\Documents\Test_, and even C:\Test, same result (_LCACHE00.TMP_ gets deleted every time I launch game, which ends in error, then the file never reappears and the error persists.

All the while any file create/modify works well for every other game, no matter where said files are located, I'm at a loss as to what is going on specifically with Abuse's load sequence here.

**Comment by Gryzor1363** [comment]
Thx @Torinde , tried exactly all of that, to the letter. With yet another fresh Dosbox Staging portable package and a reset conf file. To no avail. Error is persisting no, matter, what, I , do. No matter Dosbox location, or Abuse folder location, total control given at OS level for both Abuse executable and Dosbox, for all windows users and admin, read-only removed for the entirety of the folder and parent folders when applicable, error remains. LCACHE00.TPM gets deleted upon game initialization, then the error message and voila. 
I guess I'll have to fall back to Dosbox 0.80 or 0.81 for that singular, particular game. Every other game, 73 of them, work perfectly and read/write/create their files as they wish, and all reside in the same location as my inital Abuse folder. This is surreal to me.

Thanks for all the help though.

**Comment by weirddan455** [comment]
@Gryzor1363 You keep saying that LCACHE00.TMP gets deleted but where is the file coming from in the first place?  Are you not making a new clean DOS root without that file present?

I looked into this more and I think there may be a real bug in Staging.  It's just tricky since I can't reproduce it.  I added some log statements and this is what the game is doing with that file.

1. Unlink (delete) LCACHE00.TMP
2. Open LCACHE00.TMP (this fails on my system since it was just deleted)
3. Create LCACHE00.TMP
4. Close LCACHE00.TMP
5. Open LCACHE00.TMP

I think we may have a bug where deleting a file works differently on Windows than it does in POSIX.  In POSIX, you can unlink a file and the file entry is immediately deleted even if other processes have it open (the file itself will still exist until all handles have been closed but it's accessible only by file descriptor at that point).  You can then create a new file with the same name and both versions can exist simultaneously.

On Windows, this is not the case.  From https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-deletefilea

>  The DeleteFile function marks a file for deletion on close. Therefore, the file deletion does not occur until the last handle to the file is closed. Subsequent calls to [CreateFile](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) to open the file fail with ERROR_ACCESS_DENIED.
>
>  The use of POSIX delete causes the file to be deleted while handles remain open. Subsequent calls to [CreateFile](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) to open the file fail with ERROR_FILE_NOT_FOUND.

I think what may be happening on your system is some pesky anti-virus or who knows what (it's Windows so you may have lots of processes poking a files behind your back) has opened that file and delaying the delete.  The game is deleting and immediately trying to re-create so that fails.

If I'm right, you should be able to get it to work by ensuring you have a clean root with that file never present.

Either way though, this is an edge case we need to handle properly on Windows (somehow).

**Comment by johnnovak** [comment]
Can you try temporarily disabling all antivirus and backup processes @Gryzor1363, disconnect the network, do a clean reboot, and try the game again? It's ridiculous but it might give us some hints if that fixes it.

**Comment by Gryzor1363** [comment]
@weirddan455 I wrongly assumed that attempting to load the game both with, and without the LCACHE file was implicit, I stand corrected.

So to clarify, yes :

-I tried launching the game with the existing file ==> leads to the error above.
-I also tried without it ==> same error.
-Also, I tried the case when I try to run the game twice in a row, with the file present ==> error and deletion of LCACHE happens the first time, and then only the error and the file is never created again.
-Tried without the file with a pristine install from a CD as well (meaning of course without LCACHE ever being present there) as mentioned above ==> same result.

Thanks for the insight on descriptor management, been ages since last time I touche upon said topics, it is refreshing indeed.

All in all, in every configuration, the game never load, specifically with 0.82. Other version, I say again, all work properly no matter permissions and folder location or layout, both with and without the LCACHE file.

I am glad if my case can provide valuable data for handling such "edge" cases as you name them, so me raising this as an issue might not have been completely useless I suppose.

**Comment by Gryzor1363** [comment]
@johnnovak just saw your comment popping as I finished my post. Will absolutely do and let you know in a couple hours tops, when I'm able to do a reboot on my computer. I'll gladly furnish whatever piece of info that can help you guys address that in a larger measure, if that's a super rare and exotic though genuine possible anomaly on Windows.

**Comment by Gryzor1363** [comment]
So I deactivated all agents of my anti-virus (I am using Avast, in case that's of any use as a side info) after a reboot => everything happens unchanged from my last post reprised below.

> -Tried launching the game with the existing LCACHE file in the folder ==> leads to the error above, with the file being deleted. 
-Tried without it ==> same error. 
-Also, tried running the game more than twice in a row, with LCACHE present ==> error and deletion of the file happens the first time, and then only the error, with the file never being created again in subsequent tries. 
-Tried with a pristine install from a CD as well (meaning of course without the LCACHE ever being present there) as mentioned above ==> same result.

**Comment by johnnovak** [comment]
> So I deactivated all agents of my anti-virus (I am using Avast, in case that's of any use as a side info) after a reboot => everything happens unchanged from my last post reprised below.
> 
> > -Tried launching the game with the existing LCACHE file in the folder ==> leads to the error above, with the file being deleted.
> > -Tried without it ==> same error.
> > -Also, tried running the game more than twice in a row, with LCACHE present ==> error and deletion of the file happens the first time, and then only the error, with the file never being created again in subsequent tries.
> > -Tried with a pristine install from a CD as well (meaning of course without the LCACHE ever being present there) as mentioned above ==> same result.

Haha, this is getting comical 😆 Thanks for trying this, though!

**Comment by Gryzor1363** [comment]
Lol my pleasure John, this feels like an end of year joke that begs to be resolved next year :D My last ticket about Flashback was not near that fun, there is progress there ahah

**Comment by shermp** [comment]
Just for reference, I could not reproduce this on Windows 11.

**Comment by weirddan455** [comment]
Well I'm kind of at a loss here.  If the error is still happening at first launch when the cache file is not present, there's something more going on than a delayed delete as I previously suspected.

I'd love to know what's causing this problem and fix it but I'd really just be guessing at it without being able to reproduce it and debug it on my system.  Maybe I'll spin up a Windows 7 VM some day when I have some free time although I don't know why Windows 7 would behave differently than Windows 10 with basic file operations.

**Comment by Gryzor1363** [comment]
@weirddan455 well thanks anyway for taking the care of investigating despite Win 7 being considered obsolete for support. I'm equally bewildered at this particular behavior defying everything else on my system with similar file permissions/users/locations and a fresh install, and all my previous versions of Dosbox working just fine on the same. It's true that at first glance, it simply *should* work like the rest. But a "should" certainly does not always make a "would" in the world of code now does it :D

**Comment by Gryzor1363** [comment]
@jester-xbmc I actually did ! this, and also mounting as a directory, sadly none of them proved fruitful for my issue. thx for suggesting though, I welcome any passer-by, and good holidays to you

**Comment by Gryzor1363** [comment]
@Torinde yes, tried both. The ISO via the installer (Run via Dosbox itself) and the "flat" archive version that I just paste and try to run.

**Comment by Gryzor1363** [comment]
Happy new year everyone, and thanks for testing @Torinde  !

It would by all means point towards that issue being somehow the fruit of a mystic happening in my system behaving in ways no common system does , for some reason... I'll try re-re-downloading 0.82 and putting it in yet some other location within my user folder tree, but I'm losing hope here :(

**Comment by Gryzor1363** [comment]
Ok that's getting even more unbelievable : by principle after your post I tried to once again download the 3MB non-ISO version and ... managed to run the game. Third attempt was the right one, go figure, on the same folder as before....

Now there's something else here : still no luck on the same folder when I try the ISO-version via the install, which is in line with my repeated past experiences, the error persists identically and game won't run with the LCACHE mention. One thing I never noticed before and should mention here though : the 3 MB version does not seem to be the same game version as the ISO or other non-ISO version I found from various other abandonware sources in the past days.

Indeed, initialization console shows "Abuse 2.0" in those versions (the ISO file in the link being also 2.0), as opposed to 1.0 for the one that ended up working. Besides, screen title art, and some contrast options at the beginning are different, this is clearly not the same iteration, as I know Abuse development involved such versioning by its authors to optimize some stuff and add a better level editor down the line etc...

@Torinde , would you be so kind as to possibly try the ISO version in the link on your VM when you have time please ? since this one is Abuse 2.0, we'll see if you also get past the loading screen and if yes, we'll know once and for all that my system is unquestionably at fault , which will make that ticket a non-Dosbox related issue ... Thanks

**Comment by johnnovak** [comment]
@Gryzor1363 Sorry man, but just a tip: can you try to use the minimal number of words in these reports and only state facts? I find it difficult to extract the actual information from your posts. Please just state: tested version X, in folder X (here's the download link), with this result etc. Just a few lines is sufficient, thanks 😄

**Comment by Gryzor1363** [comment]
@johnnovak I just edited my post in kind, hope this is better ?

**Comment by johnnovak** [comment]
> @johnnovak I just edited my post in kind, hope this is better ?

Thanks man, it's much clearer now 👍🏻

**Comment by Gryzor1363** [comment]
> > @johnnovak I just edited my post in kind, hope this is better ?
> 
> Thanks man, it's much clearer now 👍🏻
> 
> PS: Happy new year to you too 😎

My pleasure mate, and thanks for the wishes :D

**Comment by Gryzor1363** [comment]
@Torinde thanks a lot for experimenting and providing such detailed feedback while offering a workaround, I appreciate the help. If this can bring anything new to the table regarding filesystem management and prevent future such occurrences in higher versions of Windows, it would make this thread worthwhile.


> I doubt @Gryzor1363 uses network share.

I confirm not relying o it for my Dosbox uses, all required files are stored locally.

> Then I mount as C an IMG file located on the network share, game works (game files located at D, which is not-good-folder-for-LCACHE; LCACHE being created in small C.IMG). So, @Gryzor1363, you can use that as a workaround...

Could you please clarify what you mean exactly by this :

- mounting the game folder as a disk image file in the IMG format in Dosbox ? (if so, I admit I did not know it was an option to begin with, that feature to me so far being only limited to floppy and CD images)
- creating such image outside of Dosbox, then mounting it as such within it ?

With that said, don't bother yourself too much anyways, for now fallbacking on 0.80 or 81 is a viable option to me as it stands, using a multi-configuration GUI makes it seamless and a clic away.

**Comment by weirddan455** [comment]
Thanks for testing and reproducing the bug @Torinde Let me just summarize the last couple posts and make sure I'm understanding.

Windows 10:
- Works on local drives
- Works on FAT image
- Broken on network shares (regression from 0.81.0)

Windows 7:
- Works with FAT image
- Broken on local drives and network shares (regression from 0.81.0)

It's probably going to be a regression from when I moved file I/O from `fopen` and friends to native Win32 API (`CreateFileW` and such).  I'll just need to get a network share or Win7 VM setup to figure out what went wrong exactly.

Not surprising that FAT images continue to work.  I didn't touch much in that part of the code and it doesn't use OS level file I/O anyway apart from opening the image itself.

**Comment by weirddan455** [comment]
I was able to reproduce the problem on Windows 10 using a network share (technically still a local folder in my case, I just shared it and mounted it in DOSBox with the network name).

Root problem is buggy game code.  I didn't look far enough at my logs when I was logging what this game was doing a few days ago.  After it does its initial delete + recreate + open, it then deletes the file a second time while it's still holding a file handle open.  In real DOS, I think this is just undefined behavior.

Apparently in Windows behavior differs based on local or network and what version of OS you're using (because it's Windows so of course everything is janky 😆).  I actually thought this should fail everywhere but for some reason in Windows 10 and newer (and with a local NTFS drive), the file can be deleted and re-created even though there's still an open handle.

I'm not sure if there's a good solution to this on the DOSBox side.  We could try and close any open file handles before an unlink but if a game tries to re-use that handle we closed behind its back, it could read/write to a completely different file (because file handles can get re-used after they are closed).

I did find a workaround though.  First time I found a useful answer on Microsoft answers:  https://answers.microsoft.com/en-us/windows/forum/all/protect-files-from-being-deleted/814dbe4c-8957-4e39-9fd7-b613e2812851

I had to add these Deny permissions to the root DOS folder.  It just prevents files from being deleted.  The delete isn't needed in this case because `CreateFile` will truncate the file to zero if it exists.

![deny-delete](https://github.com/user-attachments/assets/8e0a5729-840f-428f-8e9d-393043fbd17b)

**Comment by johnnovak** [comment]
> I'm not sure if there's a good solution to this on the DOSBox side. We could try and close any open file handles before an unlink but if a game tries to re-use that handle we closed behind its back, it could read/write to a completely different file (because file handles can get re-used after they are closed).

Can you elaborate step by step what you think would be happening in this case? I read this sentence like 5 times but I'm not sure where you're talking about native file handles, and where the emulated DOS file handles. I just don't see how this "reuse" could be happening. In my mind, we can 100% solve this in our native filesystem access code: just close the native file handle, delete the file, and that's it. Even in real DOS, once the file is deleted, the DOS file handle is orphaned and it couldn't possibly do anything useful anymore once the file is gone, no?

**Comment by weirddan455** [comment]
>  Can you elaborate step by step what you think would be happening in this case?

This is how we store DOS file handles: `std::array<std::unique_ptr<DOS_File>, DOS_FILES> Files = {};`.  Just an array of 255 pointers.  In the case of local files, the native file handle is inside that `DOS_File` class.  So if we only close the native file handle but leave the DOS handle alone, it's now pointing to an invalid native file handle which I know can be re-used on POSIX (it's just an integer and implementations often just pick the lowest free one).  I believe Windows also uses integers but unsure how they work in practice.  I assume they can also be re-used.

If we close everything including the DOS file handle, we run into the same problem on the DOS side.  DOS file handles end up as indices into our array.  When we close a file, we reset the `unique_ptr` to null.  When we open a file, we find the first null entry, create a `DOS_File`, and return the index (oversimplifying a bit.  We also set a PSP or FCB entry but that's the gist).

>  PS: Wait, are you saying the delete should not happen in DOS if the file is still open?

It *shouldn't* happen but it's programmer error if it does (as is the case with this game).  I think in real DOS the delete succeeds either way but future reads and writes to the deleted file are undefined.  It probably just writes to the place on disk where the file was even if some other file got re-created there.

Would need to test on real MS-DOS to confirm behavior exactly but it would be tricky to emulate this exactly since we're dealing with host file systems.  For FAT images, we can do more accurate emulation.

**Comment by weirddan455** [comment]
Also, I should clarify the actual problem.  The delete itself succeeds in this game (and on the host Windows system).  In Windows, a delete sometimes gets delayed until all open handles are closed.  The error happens because the game does a `CreateFile` immediately following the `DeleteFile` (with file handle open).

The `CreateFile` fails because the file name is unavailable to be created until all open handles from the previous `DeleteFile` have been closed.

On POSIX, this is not a problem because it uses unlink which will allow a new file with the same name to be created immediately.  The old open file handles still point to the old data.  It just no longer has a link to the file system hierarchy.

**Comment by johnnovak** [comment]
Thanks for the explanation @weirddan455. I get it that "re-using" host file handles can be problematic, but we can always ensure no re-use happens with some more elaborate logic, flags, etc. in the host -> DOS file mapping code (talking in generics here). I mean, not saying we _should_ do that, but we could -- that was my point 😄. But okay, I get the limitations of the current simple approach.

> The error happens because the game does a `CreateFile` immediately after the `DeleteFile` (with file handle open).

But it worked with regular fopen/fclose/remove or whatever we used before instead of the native Windows calls, right?

You said:

> It's probably going to be a regression from when I moved file I/O from fopen and friends to native Win32 API (CreateFileW and such). I'll just need to get a network share or Win7 VM setup to figure out what went wrong exactly.

So, to me the question is: what are the C stdlib file I/O calls doing differently? First I thought `DeleteFile` is asynchronous, but it's not... or not really 😄 

>  The operations Copy­File and Delete­File are synchronous. However, the NT model for file deletion is that a file is deleted when the last open handle is closed.¹ If Delete­File returns and the file still exists, then it means that somebody else still has an open handle to the file.
>
> So who has the open handle? The file was freshly created, so there can’t be any pre-existing handles to the file, and we never open it between the copy and the delete.
>
> My psychic powers said, “The offending component is your anti-virus software.”
>
> I can think of two types of software that goes around snooping on recently-created files. One of them is an indexing tool, but those tend not to be very aggressive about accessing files the moment they are created. They tend to wait until the computer is idle to do their work. Anti-virus software, however, runs in real-time mode, where they check every file as it is created. And that’s more likely to be the software that snuck in and opened the file after the copy completes so it can perform a scan on it, and that open is the extra handle that is preventing the deletion from completing. 

Anyway, the whole thing is worth reading:
https://devblogs.microsoft.com/oldnewthing/20120907-00/?p=6663

So can't we mimic in our implementation whatever the C file I/O calls are doing? Maybe a short blocking wait (with a timeout) after DeleteFile to ensure the file is really delete (maybe just 10-20 ms, so best-effort, but it should hopefully help in most cases).

**Comment by weirddan455** [comment]
>  So, to me the question is: what are the C stdlib file I/O calls doing differently?

I would have to dig into the Windows C library code to answer that.  But maybe there's something simple we can do differently on the Win32 API side.

>  Maybe a short blocking wait (with a timeout, of course) after DeleteFile to ensure the file is really gone?

That doesn't help.  It's the game itself (or DOSBox on the game's behalf) that's holding the file handle open.  We're not talking about third-party processes here (though that's also a separate concern).  We have to close the file handle ourselves if we want it closed.

>  Again, we could complicate the native to DOS file mapping code to flag that when this happens, we close all open native handles for that file first and proceed with the physical delete, then mark all open DOS handles with a flag. If that flag is set, all subsequent operations on the flagged and still-open DOS handles would be simply no-ops. That would also work, no?

Yeah, I think that can work.  It's a lot of extra logic to handle this edge case but I'm not opposed to doing something like this.

>  All I'm saying, nobody says there should be a strict 1:1 mapping between DOS and native file operations; we've been able to get away with that so far, but there's no "rule" it must be so.

Agree.  It's kind of done the way it is out of simplicity and it's worked pretty well so far.  We can extend it like your suggestion above though if needed.

Keep in mind that local file mounts are never going to be 100% accurate though.  It's the tradeoff for making it very convenient to store DOS files on a modern system.  The more accurate option is FAT images where we can control the low level filesystem implementation ourselves without host filesystem details getting in the way.

**Comment by johnnovak** [comment]
> Yeah, I think that can work. It's a lot of extra logic to handle this edge case but I'm not opposed to doing something like this.

Technically, we'll need to either do this or some other mitigation strategy to fix this regression. This is a legitimate regression that prevents people from playing the game on Windows.

> Keep in mind that local file mounts are never going to be 100% accurate though. It's the tradeoff for making it very convenient to store DOS files on a modern system. The more accurate option is FAT images where we can control the low level filesystem implementation ourselves without host filesystem details getting in the way.

Yeah, I fully understand that. 100% FAT compatibility is not even needed, unless you want to be able to run disk partitioning and undelete programs (way out of scope for us). So although that's the "ultimate" solution, it's very inconvenient and it has a lot of drawbacks, so we don't need to go there.

But making our host filesystem based solution more accurately reflect the DOS side would move our host FS solution from 95% accuracy to 98% or something, so IMO worthwhile.

I guess we could also "deal" with this regression by saying "well, you just gotta run Abuse from a HDD image". If the workaround is super difficult to do, yeah we can do that, but it would not be my preference.

**Comment by weirddan455** [comment]
>  TL;DR is we broke Abuse in 0.82.0 in Windows 10 (and beyond) only (it's fine on Linux, macOS, and Windows 7, but that doesn't help you 😄).

Other way around.  It works on Windows 10 (and Windows 11 according to @shermp) unless you're using a network drive.  Windows 7 is broken all around.  Linux is not affected and Mac shouldn't be either.

**Comment by shermp** [comment]
@weirddan455 I'm having a very quick look, so I'm probably missing something, but having a look in `drive_local.cpp`, files are created with `create_native_file()`, but unlinked using a call to `remove()`. Should they also be unlinked using OS specific calls?

**Comment by weirddan455** [comment]
@shermp We might want to use native calls instead of `remove()`.  I tested with `DeleteFile` already though and it didn't change anything.

I just put a draft PR in #4142 that fixes this game by changing the sharing flags.  I still want to do some more investigation on a couple things before merge.  I'll try to do that tomorrow.

**Comment by johnnovak** [comment]
> > TL;DR is we broke Abuse in 0.82.0 in Windows 10 (and beyond) only (it's fine on Linux, macOS, and Windows 7, but that doesn't help you 😄).
> 
> Other way around. It works on Windows 10 (and Windows 11 according to @shermp) unless you're using a network drive. Windows 7 is broken all around. Linux is not affected and Mac shouldn't be either.

Sorry, it does my head in 🤯

**Comment by Gryzor1363** [comment]
Holy cow, the day I expected to trigger such a storm possibly involving a refresh of the low-level I/O management code of Staging !!

> Thanks again @Gryzor1363 for reporting this, this has been most valuable for all of us; we've all learned something 😆

@johnnovak LOL well thanks to you for suggesting tentative quickfixes and also all other participants for indulging in even considering and testing in detail something seemingly essentially relevant to an outdated and now officially unsupported OS.

I also appreciate the vulgarized follow-up of which I read every single entry, fascinating stuff that led to me to open my 2006 PDF course on POSIX dating back to my Uni classes on descriptors and the like in my (very amateur) C days :D Besides there is already a prospective code update by @weirddan455 that so far proved very promising.

The whole dedication reminds me I absolutely have to start donating to your fantastic piece of software. 

Now I will just sit back, stop spamming here and let the pros do their magic ;-)

**Comment by weirddan455** [comment]
>  Even though obvious, I am not sure this has been explicitly confirmed in this thread but this workaround resolves my case relying on local files as well and not a shared drive. So thanks for that @weirddan455

Good to know.  I hadn't tested that on Windows 7.  Windows 10 doesn't need the workaround for local drives.

>  stems from the original dev code error or rather lack of strict filesystem-level checks around creates and deletes, leaving room for error and relying entirely to the host OS for collection of orphaned files pointers

Basically, yes.  DOS has a lot less safety checks than modern OSs do.  It kind of got away with it due to being mostly single task.  Once OSs went to true multi-tasking, they needed to more strictly define what happens when multiple file handles are held open and automatically clean things up when a process exits.

I still need to test how real DOS handles this specific case and try emulate it as best as possible across the board.  Windows and POSIX have very different models for file I/O and this discrepancy caused a bug in Windows in this case.

The game is doing buggy things but we still need to handle that as close as real DOS would if possible.  Crystal Caves had a similar issue we fixed a while back.  I think it also leaks file handles.  #3262

**Comment by weirddan455** [comment]
@Gryzor1363 This should be fixed by #4152 now

I don't have a Windows 7 machine to test on.  I tested it with the Windows 10 network drive case which I'm pretty sure is the same root cause.

If you'd like to grab a dev build and test for yourself that would be good:  https://www.dosbox-staging.org/releases/development-builds/

It's a little hard to find the actual download link.  This one should work:  https://github.com/dosbox-staging/dosbox-staging/actions/runs/12754109409/artifacts/2424153626

**Comment by Gryzor1363** [comment]
Hi @weirddan455 thanks for notifying, will test the dev build tomorrow at the latest, that was fast !

Just so I know, it is a drop-in update that I can use "as is" in place of the existing 0.82 release exe, or should I first make sure to generate a new config file like we do in major updates ?


Thanks again

**Comment by weirddan455** [comment]
>  Just so I know, it is a drop-in update that I can use "as is" in place of the existing 0.82 release exe, or should I first make sure to generate a new config file like we do in major updates ?

Depends on if you want to use it for more than a quick test.  There's been some new config options added since 0.82.0 (ex. for SoundCanvas support).  The fix for Abuse does not rely on any config options so your old config should work.

**Comment by Gryzor1363** [comment]
@weirddan455 , I can confirm that alpha build works with standard permissions on any folder, just like it did in previous versions. 

Tested with C:\Games\Classics\Abuse with custom permissions reset

- works with 0.83 alpha
- is broken again with 0.82, triggers LCACHE error

Test with freshly created C:\Games\Classics\Test :

- works with 0.83 alpha
- triggers LCACHE error with 0.82


Again thanks for considering ticket and also the many replies and valuable insight on the underlying topics. 
Looks like Dosbox still has good days ahead of itself on good old Windows 7  after all :)

A good end of week to all.


PS : Should I close myself or leave that to you (possibly with the right comments) ?

**Comment by weirddan455** [comment]
Thanks for testing @Gryzor1363 

Closing this as solved 😄
